src/FloatingMenuPlugin.ts

```diff
@@ -0,0 +1,570 @@
+
+  return sliceModel;
+}
+
+export function copySelectionPlain(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  if (!view.hasFocus()) {
+    view.focus();
+  }
+  const { from, to } = view.state.selection;
+  if (from === to) return;
+
+  const slice = view.state.doc.slice(from, to);
+  const text = slice.content.textBetween(0, slice.content.size, '\n');
+
+  navigator.clipboard
+    .writeText(text)
+    .then(() => { }) //console.log('Plain text copied!'))
+    .catch((err) => console.error('Clipboard write failed:', err));
+  if (plugin._popUpHandle?.close) {
+    plugin._popUpHandle.close(null);
+    plugin._popUpHandle = null;
+  }
+}
+
+export async function pasteFromClipboard(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  try {
+    if (!view.hasFocus()) view.focus();
+
+    const text = await navigator.clipboard.readText();
+    let tr: Transaction;
+
+    try {
+      // Try parsing as JSON slice
+      const parsed = JSON.parse(text);
+      const slice = Slice.fromJSON(view.state.schema, parsed);
+      tr = view.state.tr.replaceSelection(slice);
+    } catch (jsonErr) {
+      // If not JSON, treat as plain text
+      tr = view.state.tr.insertText(
+        text,
+        view.state.selection.from,
+        view.state.selection.to
+      );
+    }
+
+    view.dispatch(tr.scrollIntoView());
+  } catch (err) {
+    console.error('Clipboard paste failed:', err);
+  } finally {
+    if (plugin._popUpHandle?.close) {
+      plugin._popUpHandle.close(null);
+      plugin._popUpHandle = null;
+    }
+  }
+}
+
+export async function pasteAsReference(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  try {
+    if (!view.hasFocus()) view.focus();
+    const text = await navigator.clipboard.readText();
+    const parsed = JSON.parse(text);
+    const sliceModel: SliceModel = parsed.sliceModel;
+
+    if (!plugin.sliceManager?.createSliceViaDialog) {
+      throw new Error(
+        'SliceManager or createSliceViaDialog is not initialized'
+      );
+    }
+
+    const val = await plugin.sliceManager.createSliceViaDialog(sliceModel);
+    if (!val) {
+      return;
+    }
+    insertReference(
+      view,
+      val.id,
+      val.source,
+      view['docView']?.node?.attrs?.objectMetaData?.name
+    );
+  } catch (err) {
+    console.error('Failed to paste content or create slice:', err);
+  } finally {
+    if (plugin._popUpHandle?.close) {
+      plugin._popUpHandle.close(null);
+      plugin._popUpHandle = null;
+    }
+  }
+}
+
+export async function pasteAsPlainText(
+  view: EditorView,
+  plugin: FloatingMenuPlugin
+) {
+  try {
+    if (!view.hasFocus()) view.focus();
+
+    const text = await navigator.clipboard.readText();
+    let plainText = text;
+
+    try {
+      const parsed = JSON.parse(text);
+      const slice = Slice.fromJSON(view.state.schema, parsed);
+
+      const frag = slice.content;
+      plainText = '';
+      frag.forEach((node) => {
```

Comment:
Sonar: Use `for…of` instead of `.forEach(…)`.

You can NOSONAR this rule for Sonar fragments until Sonar fixes this typing bug.
```suggestion
      frag.forEach((node) => { // NOSONAR not an iterable
```